### ISSUE TITLE
HTML printing using 2.1's API (sample.html from 2.1) is blurry when using QZ-Tray 2.0.11

### ISSUE THREAD

**Issue by lite1979** [qzind/tray#594]
When printing Pixel> HTML to a Star TSP143 thermal receipt printer, HTML prints good quality from 2.0.11's sample.html when using QZ-Tray 2.0.11

It also prints well from 2.1.0's sample.html when using QZ-Tray 2.1.0+1

When using 2.1.0's sample.html, in combination with QZ-Tray 2.0.11, however, an acceptable print could not be achieved, despite trying different config options.

Good result configs were separated by: 
![QZmixedversiontest](https://user-images.githubusercontent.com/3408749/75284626-2816d880-57e3-11ea-8330-986efbd68bde.jpg)

2.0.11: rasterize: true

2.1.0+1 rasterize: false

Other config options remained the same for the good prints: 
color type: color, 
scaleContent: true, 
orientation: default, 
density: blank, 
rotation: 0

**Comment by lite1979** [comment]
![QZmixedversionhtmltest](https://user-images.githubusercontent.com/3408749/75277580-432f1b80-57d6-11ea-968d-766981148275.jpg)

**Comment by stewartbrad** [comment]
Thanks Tres

**Comment by tresf** [comment]
Interestingly, I cannot reproduce this problem using sample.html from 2.0 versus sample.html from 2.1, they both yield very bad results, leading to a dead end.  I'm a bit curious how good the quality is with 2.0 to begin with, I have a feeling that this is a deeper-rooted bug than simply the switch to 2.1.  I'll post my findings ASAP.

**Comment by tresf** [comment]
Getting a bit closer... here's the difference between good and bad prints... 

What's happening is when printing using the 2.1 demo page, the higher density is lost.

```diff
[TRACE] 2020-02-25 16:51:32,806 @ qz.printer.action.WebApp:?
	Done: 0.0 > 100.0
[TRACE] 2020-02-25 16:51:32,812 @ qz.printer.action.WebApp:?
	New state: RUNNING > SUCCEEDED
[TRACE] 2020-02-25 16:51:32,820 @ qz.printer.action.WebApp:?
-	Zooming in by x1.0 for increased quality
+	Zooming in by x2.5 for increased quality
[TRACE] 2020-02-25 16:51:32,822 @ qz.printer.action.WebApp:?
-	Setting HTML page width to 754.56
+	Setting HTML page width to 301.82399999999996
[TRACE] 2020-02-25 16:51:32,932 @ qz.printer.action.WebApp:?
-	Setting HTML page height to 672.0
+	Setting HTML page height to 1655.0
[DEBUG] 2020-02-25 16:51:33,044 @ qz.printer.action.WebApp:?
	Attempting image capture
[TRACE] 2020-02-25 16:51:33,697 @ qz.printer.action.PrintPixel:?
-	DPI: 0.0	CNV: 72.0
+	DPI: 180.0	CNV: 72.0
[TRACE] 2020-02-25 16:51:33,698 @ qz.printer.action.PrintPixel:?
	Drawable area: 0.0,0.0:3.144,11.688999
[TRACE] 2020-02-25 16:51:33,698 @ qz.printer.action.PrintPixel:?
-	[color, (0.0,0.0)->(79.858,296.901)mm]
+	[18000x18000 dphi, color, (0.0,0.0)->(79.858,296.901)mm]

**Comment by stewartbrad** [comment]
> Getting a bit closer... here's the difference between good and bad prints...
> 
> What's happening is when printing using the 2.1 demo page, the higher density is lost.
> 
> ```diff
> [TRACE] 2020-02-25 16:51:32,806 @ qz.printer.action.WebApp:?
> 	Done: 0.0 > 100.0
> [TRACE] 2020-02-25 16:51:32,812 @ qz.printer.action.WebApp:?
> 	New state: RUNNING > SUCCEEDED
> [TRACE] 2020-02-25 16:51:32,820 @ qz.printer.action.WebApp:?
> -	Zooming in by x1.0 for increased quality
> +	Zooming in by x2.5 for increased quality
> [TRACE] 2020-02-25 16:51:32,822 @ qz.printer.action.WebApp:?
> -	Setting HTML page width to 754.56
> +	Setting HTML page width to 301.82399999999996
> [TRACE] 2020-02-25 16:51:32,932 @ qz.printer.action.WebApp:?
> -	Setting HTML page height to 672.0
> +	Setting HTML page height to 1655.0
> [DEBUG] 2020-02-25 16:51:33,044 @ qz.printer.action.WebApp:?
> 	Attempting image capture
> [TRACE] 2020-02-25 16:51:33,697 @ qz.printer.action.PrintPixel:?
> -	DPI: 0.0	CNV: 72.0
> +	DPI: 180.0	CNV: 72.0
> [TRACE] 2020-02-25 16:51:33,698 @ qz.printer.action.PrintPixel:?
> 	Drawable area: 0.0,0.0:3.144,11.688999
> [TRACE] 2020-02-25 16:51:33,698 @ qz.printer.action.PrintPixel:?
> -	[color, (0.0,0.0)->(79.858,296.901)mm]
> +	[18000x18000 dphi, color, (0.0,0.0)->(79.858,296.901)mm]
> ```

interesting...
i currently have a bunch of potential DPI values i put into the density array, but i couldn't tell which ones made a difference, whether i went above the TSP143's 203 DPI, below, used fallback density, used no densities, etc.

by the way just to follow along, in terms of the demo page version you're using, when you say 2.1 demo page that's the same as me using the 2.1 npm package of QZ on my end, correct?

**Comment by tresf** [comment]
> 2.1 demo page that's the same as me using the 2.1 npm package of QZ on my end, correct?

Correct.

So I think I know what's going on... in 2.0 there was only ever one value for `rasterize: ` and it was `rasterize: true`.   There's really no way to NOT rasterize with 2.0 because we didn't offer a vector printing technique back then.

The real issue is that `rasterize: false` is simply undefined and unsupported behavior.  The closest bug report we have on record for this is #257 (ignore the title but read @bberenz's commit about "Always treating images as rasterize".  This smells as an identical bug, but for the HTML printing.  The reason this is rarely exposed is because 2.0 defaulted to `false` directly inside the API whereas 2.1 defaults to 2.1, so the value is changing without warning.

2.0:
https://github.com/qzind/tray/blob/066b0c6d31edba72da6fd6625cd88387d5c844fd/js/qz-tray.js#L398

2.1:
https://github.com/qzind/tray/blob/34346d399dac36a24c1aef3ea5c007f02dd0c716/js/qz-tray.js#L419

The workaround is to simply detect and set `rasterize: true` when using qz tray <= 2.0.  We should really handle this in the compat function though... 

https://github.com/qzind/tray/blob/34346d399dac36a24c1aef3ea5c007f02dd0c716/js/qz-tray.js#L665-L677

... but at time of writing this, we don't touch the config, so that's something I'll ask @bberenz to look into... 

@stewartbrad can you try this:

```js
var config = qz.configs.create("Star TSP143", { /* whatever options you set normally */ });
if(qz.api.isVersion(2.0)) {
    // Workaround for https://github.com/qzind/tray/issues/594
    qz.reconfigure({ rasterize: true });
}

```

@bberenz can we add a way to set rasterize to true if:
* It's HTML<br>-- AND --
* It's not explicitly set<br>-- AND -- 
* The version of QZ Tray is 2.0?

**Comment by tresf** [comment]
> i currently have a bunch of potential DPI values i put into the density array, but i couldn't tell which ones made a difference, whether i went above the TSP143's 203 DPI, below, used fallback density, used no densities, etc.

As an aside, I strongly encourage you to avoid specifying `{ density: ...}` in your config, at least on Windows.  In Windows -- at least for my Epson -- the DPI is reported properly and we'll trust what the driver supports which will prevent strangenesses for printers that have different DPIs.

We do have a parameter called `{ fallbackDensity: }` which will allow systems that don't report the density properly to have something to fallback on (e.g. macOS, Linux).

**Comment by stewartbrad** [comment]
> > i currently have a bunch of potential DPI values i put into the density array, but i couldn't tell which ones made a difference, whether i went above the TSP143's 203 DPI, below, used fallback density, used no densities, etc.
> 
> As an aside, I strongly encourage you to avoid specifying `{ density: ...}` in your config, at least on Windows. In Windows -- at least for my Epson -- the DPI is reported properly and we'll trust what the driver supports which will prevent strangenesses for printers that have different DPIs (as an example, my Epson is 180 dpi, not 203dpi!)
> 
> We do have a parameter called `{ fallbackDensity: ... }` which will allow systems that don't report the density properly to have something to fallback on (e.g. macOS, Linux).

ok thanks, i was actually going to ask you about that

i wasn't sure how the DPI reporting worked between the OS, printer and QZ.  we have clients using all kinds of different things (1 guy was even using his inkjet), so it would be great to let the system handle that value.

i'll remove density from my config now.
question... if fallbackDensity is also set on Windows, would that override the detected density from the printer? i can conditionally set fallback or not based on platform detection if necessary

**Comment by tresf** [comment]
> if `fallbackDensity` is also set on Windows, would that override the detected density from the printer? i can conditionally set fallback or not based on platform detection if necessary

If the documentation and my memory serves correctly, it's only as a fallback for when it can't be detected using the Java API.

**Comment by stewartbrad** [comment]
> > if `fallbackDensity` is also set on Windows, would that override the detected density from the printer? i can conditionally set fallback or not based on platform detection if necessary
> 
> If the documentation and my memory serves correctly, it's only as a fallback for when it can't be detected using the Java API.

alright. i threw the condition in anyway just to be safe

wanted to run this by you, see if there's any parameters you think would be worth adding / removing?

`        var config = qz.configs.create(printername, {
          copies: 1,
          colorType: 'color',
          scaleContent: true,
          rasterize: true,
          ...(!Platform.is.win ? {fallbackDensity: 203} : {}),
          size: { width: 2.83 }
        })`

this is my temporary "2.0.11-compatible" config, will update it as things are figured out.

i have that width set just to keep receipts capped at the standard 72mm width.  last month one of our older test printers (HiSpeed POS-X) was being problematic with scaling and page width, currently printing fine but i'm not sure which particular thing fixed it.

main goal is to reach a point where everything fits with decent quality on any normal receipt printer (out of the box, no changes to printer properties or preferences)

**Comment by bberenz** [comment]
A patch to the 2.1 js for compatibility with 2.0 config defaults is ready for review on the `added-compat` branch, via 9fd2ef2e6967ff5ef6701dc944397fd3096ee05a. 

The 2.0 java change is directly available on the `2.0` branch via d0b320cb80d043f6da2c64c2b56adca76def7220, which just forces rasterize for all html prints the same as images.

**Comment by stewartbrad** [comment]
> A patch to the 2.1 js for compatibility with 2.0 config defaults is ready for review on the `added-compat` branch, via [9fd2ef2](https://github.com/qzind/tray/commit/9fd2ef2e6967ff5ef6701dc944397fd3096ee05a).
> 
> The 2.0 java change is directly available on the `2.0` branch via [d0b320c](https://github.com/qzind/tray/commit/d0b320cb80d043f6da2c64c2b56adca76def7220), which just forces rasterize for all html prints the same as images.

Thanks Brett.

I changed my code back to use just the 1 config then tested out 2.0.11 and 2.1.0+1 with your latest from `added-compat` 

It handles both now which is great (pic attached).  The 2.0.11 looks about equal to where i had it (with all the specifically set config params), but now it appears to handle all that as default. 
slightly less precision than the 2.1.0+1, but it might have always been like that in our pre-2.1 versions, not a big deal.

lemme know what you think.  and thanks again for the fast update
![receipts](https://user-images.githubusercontent.com/61476202/75320000-b9b23480-583b-11ea-9748-6124337ecb98.jpg)

**Comment by tresf** [comment]
> wanted to run this by you, see if there's any parameters you think would be worth adding / removing?
> ```
>    var config
>    if (
>        String(qzgetversion) === '2.1' ||
>        String(qzgetversion) === '2.1.0' ||
>        String(qzgetversion) === '2.1.0+1'
>        ) {
>          config = qz.configs.create(printername, {
>            copies: 1,
>            size: { width: 2.83 }
>          })
>    } else {
>     config = qz.configs.create(printername, {
>        copies: 1,
>        colorType: 'color',
>        scaleContent: true,
>        rasterize: true,
>        ...(!Platform.is.win ? {fallbackDensity: 203} : {}),
>        size: { width: 2.83 }
>      })
>    }
>   ```

Sure thing.

* 🚫 `copies` can be removed, the api always defaults to `1`
* 🚫 `colorType` can be removed.  I know Dymo printers prefer `grayscale|blackwhite` but `color` is the default in the API.
* ✅ `scaleContent: true` needs to be provided to stretch to different paper sizes
* ⚠️ `size` can probably be removed, depending on if the default paper is setup in the driver every time.
* ⚠️ `fallbackDensity`  can be provided on all prints, not just Mac/Linux.  It won't hurt Windows, it's just a fallback and might help if a 3rd party Windows driver doesn't offer density.
* ⚠️ `rasterize: true` depends on which qz-tray.js version (e.g. Brett's patched version avoids the need for this)

Disclaimer, although I recommend taking these away if we change any defaults down the road, it never hurts to manually specify them so stuff like this exact bug doesn't occur, so the choice is up to you.

Last, for simplicity you can always use `qz.api.isVersion(2.0)` to simplify the multiple conditionals.  Note, this is new with qz-tray.js 2.1.0.

**Comment by stewartbrad** [comment]
> > wanted to run this by you, see if there's any parameters you think would be worth adding / removing?
> > ```
> >    var config
> >    if (
> >        String(qzgetversion) === '2.1' ||
> >        String(qzgetversion) === '2.1.0' ||
> >        String(qzgetversion) === '2.1.0+1'
> >        ) {
> >          config = qz.configs.create(printername, {
> >            copies: 1,
> >            size: { width: 2.83 }
> >          })
> >    } else {
> >     config = qz.configs.create(printername, {
> >        copies: 1,
> >        colorType: 'color',
> >        scaleContent: true,
> >        rasterize: true,
> >        ...(!Platform.is.win ? {fallbackDensity: 203} : {}),
> >        size: { width: 2.83 }
> >      })
> >    }
> > ```
> 
> Sure thing.
> 
> * 🚫 `copies` can be removed, the api always defaults to `1`
> * 🚫 `colorType` can be removed.  I know Dymo printers prefer `grayscale|blackwhite` but `color` is the default in the API.
> * ✅ `scaleContent: true` keep.  Needs to be provided to stretch to different paper sizes
> * ⚠️ `size` can probably be removed, depending on if the default paper is setup in the driver every time.
> * ⚠️ `fallbackDensity`  can be provided on all prints, not just Mac/Linux.  It won't hurt Windows, it's just a fallback and might help if a 3rd party Windows driver doesn't offer density.
> * ⚠️ `rasterize: true` depends on which qz-tray.js version (e.g. Brett's patched version avoids the need for this)
> 
> Disclaimer, although I recommend taking these away if we change any defaults down the road, it never hurts to manually specify them so stuff like this exact bug doesn't occur, so the choice is up to you.
> 
> Last, for simplicity you can always use `qz.api.isVersion(2.0)` to simplify the multiple conditionals. Note, this is new with qz-tray.js 2.1.0.

thanks.  i implemented these suggestions.

we're about to release our next beta.  previously we had been directing them to your /download page for printer setup, but I've changed that link now to go here - https://github.com/qzind/tray/releases/tag/v2.1.0 (because we want new users to start with 2.1.x)
any potential issues with us doing that?

also, i had previously included a link to adoptOpenJDK (this one - https://adoptopenjdk.net/?variant=openjdk11&jvmVariant=hotspot), but I'm considering changing that as well.
should we be directing users to a different version of Java?  maybe one with JavaFX bundled if they're also installing 2.1?
or should we omit the java link, if the QZ install wizard already directs you to java (if it's not on your system already)?
what do you think the best approach is?

**Comment by tresf** [comment]
2.1 has FX bundled.

AdoptOpenJDK11 HotSpot per the exact link you shared is the recommended version at time of writing this.  We'll be bundling it as soon as possible, hopefully starting with 2.2.0.

The download link to the GitHub page is OK but it'll change when we release 2.1.1 and the GitHub page can be confusing to navigate.

I'll see to make a way to filter our Downloads page to make this easier on users, this will also help out "upgrade now" button we've placed in our About dialog.

**Comment by tresf** [comment]
> > Last, for simplicity you can always use `qz.api.isVersion(2.0)` to simplify the multiple conditionals. Note, this is new with qz-tray.js 2.1.0.

Correct, the API usage is: `qz.api.isVersion(2, 0)`.  The commas is needed for proper comparison.

Note, this feature is patched internally since https://github.com/qzind/tray/pull/595 (2.1) and https://github.com/qzind/tray/commit/d0b320cb80d043f6da2c64c2b56adca76def7220 (2.0).

The releases containing this fix will be 2.1.1, 2.0.12 respectively.

**Comment by stewartbrad** [comment]
> > > Last, for simplicity you can always use `qz.api.isVersion(2.0)` to simplify the multiple conditionals. Note, this is new with qz-tray.js 2.1.0.
> 
> Correction, the API usage is: `qz.api.isVersion(2, 0)`. The commas is needed for proper comparison.
> 
> Note, this feature is patched internally since #595 (2.1) and [d0b320c](https://github.com/qzind/tray/commit/d0b320cb80d043f6da2c64c2b56adca76def7220) (2.0).
> 
> The releases containing this fix will be 2.1.1, 2.0.12 respectively.

instead of api.isVersion i'm using qz.api.getVersion currently, then checking for all variations of 2.1, 2.1.X and 2.1.X+X
if it's not those, then i use the older 2.0.X config

will check back in with you later today or tomorrow.  we did our global launch today, no problems with printing so far (other than people not knowing what a download is)
